invalid memory address or nil pointer dereference when updating resource

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
0.12.16
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
variable "test" {
  default = "a"
}

resource "google_compute_instance_template" "asd" {
  for_each    = signum(length(var.asd.a.servers) + length(var.asd.b.servers) + length(var.asd.c.servers) + length(var.asd.d.servers)) > 0 ? local.somethings : toset([])
  name_prefix = terraform.workspace
  tags = [
    terraform.workspace,
    "${terraform.workspace}-asd",
    "${terraform.workspace}-asd-${each.key}"
  ]

  labels = {
    something = "${each.key}"
    role      = "asd"
    workspace = terraform.workspace
  }

  machine_type = var.asd.size

  scheduling {
    automatic_restart = true
  }

  disk {
    source_image = var.test != "" ? "debian-10" : "debian-9"
    auto_delete  = true
    boot         = true
    disk_type    = "pd-ssd"
  }

  network_interface {
    network = var.network.vpc
    access_config {
      nat_ip = ""
    }
  }

  metadata = {
    something           = "${each.key}"
    role                = "asd"
    workspace           = terraform.workspace
    sshKeys             = join("\n", var.ssh_keys)
    startup-script-url  = "gs://${var.storage.scripts_bucket}/${var.storage.startup_script}"
    shutdown-script-url = "gs://${var.storage.scripts_bucket}/${var.storage.shutdown_script}"
  }

  service_account {
    scopes = [
      "storage-ro",
      "logging-write",
      "monitoring-write",
      "service-management",
      "service-control",
      "trace-append",
    ]
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "google_compute_region_instance_group_manager" "asd_c" {
  for_each = var.asd.c.servers
  name     = "${terraform.workspace}-${each.key}-asd-c"

  base_instance_name = terraform.workspace
  region             = each.key

  target_size = each.value

  update_policy {
    type                         = "PROACTIVE"
    instance_redistribution_type = "NONE"
    minimal_action               = "REPLACE"
    max_surge_fixed              = 3
    max_unavailable_fixed        = 3
    min_ready_sec                = 480
  }

  version {
    name              = "current"
    instance_template = google_compute_instance_template.asd["c"].self_link
  }
}
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1abb0a3]

goroutine 241 [running]:
github.com/hashicorp/terraform/terraform.(*evaluationStateData).getResourceInstancesAll(0xc0000c41b0, 0x4d, 0xc000044c00, 0x20, 0xc0000430a4, 0x3, 0xc0000425cc, 0x4, 0x8f, 0x19, ...)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/evaluate.go:724 +0x483
github.com/hashicorp/terraform/terraform.(*evaluationStateData).GetResource(0xc0000c41b0, 0x4d, 0xc000044c00, 0x20, 0xc0000430a4, 0x3, 0xc0000425cc, 0x4, 0x8f, 0x19, ...)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/evaluate.go:560 +0x85a
github.com/hashicorp/terraform/lang.(*Scope).evalContext(0xc00018e690, 0xc000869800, 0x6, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x100c688)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/lang/eval.go:278 +0x2d2c
github.com/hashicorp/terraform/lang.(*Scope).EvalContext(...)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/lang/eval.go:160
github.com/hashicorp/terraform/lang.(*Scope).EvalBlock(0xc00018e690, 0x2e89700, 0xc0008527b0, 0xc000251240, 0x1, 0x1, 0x0, 0x0, 0x0, 0x26b6dc0, ...)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/lang/eval.go:52 +0xf1
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).EvaluateBlock(0xc000b78680, 0x2e89700, 0xc0008527b0, 0xc000251240, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/eval_context_builtin.go:272 +0x2b4
github.com/hashicorp/terraform/terraform.(*EvalDiff).Eval(0xc000b5f200, 0x2eba080, 0xc000b78680, 0x2, 0x2, 0x0, 0x0)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/eval_diff.go:138 +0x226
github.com/hashicorp/terraform/terraform.EvalRaw(0x2e47e00, 0xc000b5f200, 0x2eba080, 0xc000b78680, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc00086cba0, 0x2eba080, 0xc000b78680, 0x2, 0x2, 0x1aa95d5, 0x2e48180)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:20 +0xfd
github.com/hashicorp/terraform/terraform.EvalRaw(0x2e48140, 0xc00086cba0, 0x2eba080, 0xc000b78680, 0x26a8240, 0x3c66b81, 0x2612b40, 0xc000811ee0)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.Eval(0x2e48140, 0xc00086cba0, 0x2eba080, 0xc000b78680, 0xc00086cba0, 0x2e48140, 0xc00086cba0, 0x0)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/eval.go:35 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x299cb60, 0xc000c465a0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/terraform/graph.go:90 +0xf7e
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000860500, 0x299cb60, 0xc000c465a0, 0xc000ca8080)
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x377
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/private/tmp/terraform-20191119-99583-jc7irh/terraform-0.12.16/src/github.com/hashicorp/terraform/dag/walk.go:314 +0xaa7
```

### Expected Behavior
<!--
What should have happened?
-->
A new template should have been created and `google_compute_region_instance_group_manager` updated accordingly

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
1. Create the resources above
2. Change variable's `test` value and `terraform apply` again

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
Plan
```hcl
+/- resource "google_compute_instance_template" "asd" {
        can_ip_forward       = false
      ~ id                   = "projects/someproject/global/instanceTemplates/default20191128160228051400000002" -> (known after apply)
        labels               = {
            "role"      = "asd"
            "something" = "a"
            "workspace" = "default"
        }
        machine_type         = "n1-highmem-8"
        metadata             = {
            "role"                = "asd"
            "shutdown-script-url" = "gs:///"
            "something"           = "a"
            "sshKeys"             = ""
            "startup-script-url"  = "gs:///"
            "workspace"           = "default"
        }
      ~ metadata_fingerprint = "h65o9TxHOc4=" -> (known after apply)
      ~ name                 = "default20191128160228051400000002" -> (known after apply)
        name_prefix          = "default"
      ~ project              = "someproject" -> (known after apply)
      + region               = (known after apply)
      ~ self_link            = "https://www.googleapis.com/compute/beta/projects/someproject/global/instanceTemplates/default20191128160228051400000002" -> (known after apply)
        tags                 = [
            "default",
            "default-asd",
            "default-asd-a",
            "default-game-service",
        ]
      + tags_fingerprint     = (known after apply)

      ~ disk {
            auto_delete  = true
            boot         = true
          ~ device_name  = "persistent-disk-0" -> (known after apply)
          - disk_size_gb = 0 -> null
            disk_type    = "pd-ssd"
          ~ interface    = "SCSI" -> (known after apply)
          - labels       = {} -> null
          ~ mode         = "READ_WRITE" -> (known after apply)
          ~ source_image = "projects/debian-cloud/global/images/family/debian-10" -> "debian-9" # forces replacement
          ~ type         = "PERSISTENT" -> (known after apply)
        }

      ~ network_interface {
          ~ network            = "https://www.googleapis.com/compute/v1/projects/someproject/global/networks/mynet" -> "mynet"
          + subnetwork         = (known after apply)
          + subnetwork_project = (known after apply)

          ~ access_config {
              + nat_ip       = (known after apply)
              ~ network_tier = "PREMIUM" -> (known after apply)
            }
        }

      ~ scheduling {
            automatic_restart   = true
          ~ on_host_maintenance = "MIGRATE" -> (known after apply)
            preemptible         = false
        }

      ~ service_account {
          ~ email  = "default" -> (known after apply)
            scopes = [
                "https://www.googleapis.com/auth/devstorage.read_only",
                "https://www.googleapis.com/auth/logging.write",
                "https://www.googleapis.com/auth/monitoring.write",
                "https://www.googleapis.com/auth/service.management.readonly",
                "https://www.googleapis.com/auth/servicecontrol",
                "https://www.googleapis.com/auth/trace.append",
            ]
        }
    }

  # google_compute_instance_template.asd["b"] must be replaced
+/- resource "google_compute_instance_template" "asd" {
        can_ip_forward       = false
      ~ id                   = "projects/someproject/global/instanceTemplates/default20191128160228076600000003" -> (known after apply)
        labels               = {
            "role"      = "asd"
            "something" = "b"
            "workspace" = "default"
        }
        machine_type         = "n1-highmem-8"
        metadata             = {
            "role"                = "asd"
            "shutdown-script-url" = "gs:///"
            "something"           = "b"
            "sshKeys"             = ""
            "startup-script-url"  = "gs:///"
            "workspace"           = "default"
        }
      ~ metadata_fingerprint = "IpDI9lytJa0=" -> (known after apply)
      ~ name                 = "default20191128160228076600000003" -> (known after apply)
        name_prefix          = "default"
      ~ project              = "someproject" -> (known after apply)
      + region               = (known after apply)
      ~ self_link            = "https://www.googleapis.com/compute/beta/projects/someproject/global/instanceTemplates/default20191128160228076600000003" -> (known after apply)
        tags                 = [
            "default",
            "default-asd",
            "default-asd-b",
            "default-game-service",
        ]
      + tags_fingerprint     = (known after apply)

      ~ disk {
            auto_delete  = true
            boot         = true
          ~ device_name  = "persistent-disk-0" -> (known after apply)
          - disk_size_gb = 0 -> null
            disk_type    = "pd-ssd"
          ~ interface    = "SCSI" -> (known after apply)
          - labels       = {} -> null
          ~ mode         = "READ_WRITE" -> (known after apply)
          ~ source_image = "projects/debian-cloud/global/images/family/debian-10" -> "debian-9" # forces replacement
          ~ type         = "PERSISTENT" -> (known after apply)
        }

      ~ network_interface {
          ~ network            = "https://www.googleapis.com/compute/v1/projects/someproject/global/networks/mynet" -> "mynet"
          + subnetwork         = (known after apply)
          + subnetwork_project = (known after apply)

          ~ access_config {
              + nat_ip       = (known after apply)
              ~ network_tier = "PREMIUM" -> (known after apply)
            }
        }

      ~ scheduling {
            automatic_restart   = true
          ~ on_host_maintenance = "MIGRATE" -> (known after apply)
            preemptible         = false
        }

      ~ service_account {
          ~ email  = "default" -> (known after apply)
            scopes = [
                "https://www.googleapis.com/auth/devstorage.read_only",
                "https://www.googleapis.com/auth/logging.write",
                "https://www.googleapis.com/auth/monitoring.write",
                "https://www.googleapis.com/auth/service.management.readonly",
                "https://www.googleapis.com/auth/servicecontrol",
                "https://www.googleapis.com/auth/trace.append",
            ]
        }
    }

  # google_compute_instance_template.asd["c"] must be replaced
+/- resource "google_compute_instance_template" "asd" {
        can_ip_forward       = false
      ~ id                   = "projects/someproject/global/instanceTemplates/default20191128160228194200000004" -> (known after apply)
        labels               = {
            "role"      = "asd"
            "something" = "c"
            "workspace" = "default"
        }
        machine_type         = "n1-highmem-8"
        metadata             = {
            "role"                = "asd"
            "shutdown-script-url" = "gs:///"
            "something"           = "c"
            "sshKeys"             = ""
            "startup-script-url"  = "gs:///"
            "workspace"           = "default"
        }
      ~ metadata_fingerprint = "5rpIRd9ihE8=" -> (known after apply)
      ~ name                 = "default20191128160228194200000004" -> (known after apply)
        name_prefix          = "default"
      ~ project              = "someproject" -> (known after apply)
      + region               = (known after apply)
      ~ self_link            = "https://www.googleapis.com/compute/beta/projects/someproject/global/instanceTemplates/default20191128160228194200000004" -> (known after apply)
        tags                 = [
            "default",
            "default-asd",
            "default-asd-c",
            "default-game-service",
        ]
      + tags_fingerprint     = (known after apply)

      ~ disk {
            auto_delete  = true
            boot         = true
          ~ device_name  = "persistent-disk-0" -> (known after apply)
          - disk_size_gb = 0 -> null
            disk_type    = "pd-ssd"
          ~ interface    = "SCSI" -> (known after apply)
          - labels       = {} -> null
          ~ mode         = "READ_WRITE" -> (known after apply)
          ~ source_image = "projects/debian-cloud/global/images/family/debian-10" -> "debian-9" # forces replacement
          ~ type         = "PERSISTENT" -> (known after apply)
        }

      ~ network_interface {
          ~ network            = "https://www.googleapis.com/compute/v1/projects/someproject/global/networks/mynet" -> "mynet"
          + subnetwork         = (known after apply)
          + subnetwork_project = (known after apply)

          ~ access_config {
              + nat_ip       = (known after apply)
              ~ network_tier = "PREMIUM" -> (known after apply)
            }
        }

      ~ scheduling {
            automatic_restart   = true
          ~ on_host_maintenance = "MIGRATE" -> (known after apply)
            preemptible         = false
        }

      ~ service_account {
          ~ email  = "default" -> (known after apply)
            scopes = [
                "https://www.googleapis.com/auth/devstorage.read_only",
                "https://www.googleapis.com/auth/logging.write",
                "https://www.googleapis.com/auth/monitoring.write",
                "https://www.googleapis.com/auth/service.management.readonly",
                "https://www.googleapis.com/auth/servicecontrol",
                "https://www.googleapis.com/auth/trace.append",
            ]
        }
    }

  # google_compute_instance_template.asd["d"] must be replaced
+/- resource "google_compute_instance_template" "asd" {
        can_ip_forward       = false
      ~ id                   = "projects/someproject/global/instanceTemplates/default20191128160228027500000001" -> (known after apply)
        labels               = {
            "role"      = "asd"
            "something" = "d"
            "workspace" = "default"
        }
        machine_type         = "n1-highmem-8"
        metadata             = {
            "role"                = "asd"
            "shutdown-script-url" = "gs:///"
            "something"           = "d"
            "sshKeys"             = ""
            "startup-script-url"  = "gs:///"
            "workspace"           = "default"
        }
      ~ metadata_fingerprint = "OZI4l7YZISU=" -> (known after apply)
      ~ name                 = "default20191128160228027500000001" -> (known after apply)
        name_prefix          = "default"
      ~ project              = "someproject" -> (known after apply)
      + region               = (known after apply)
      ~ self_link            = "https://www.googleapis.com/compute/beta/projects/someproject/global/instanceTemplates/default20191128160228027500000001" -> (known after apply)
        tags                 = [
            "default",
            "default-asd",
            "default-asd-d",
            "default-game-service",
        ]
      + tags_fingerprint     = (known after apply)

      ~ disk {
            auto_delete  = true
            boot         = true
          ~ device_name  = "persistent-disk-0" -> (known after apply)
          - disk_size_gb = 0 -> null
            disk_type    = "pd-ssd"
          ~ interface    = "SCSI" -> (known after apply)
          - labels       = {} -> null
          ~ mode         = "READ_WRITE" -> (known after apply)
          ~ source_image = "projects/debian-cloud/global/images/family/debian-10" -> "debian-9" # forces replacement
          ~ type         = "PERSISTENT" -> (known after apply)
        }

      ~ network_interface {
          ~ network            = "https://www.googleapis.com/compute/v1/projects/someproject/global/networks/mynet" -> "mynet"
          + subnetwork         = (known after apply)
          + subnetwork_project = (known after apply)

          ~ access_config {
              + nat_ip       = (known after apply)
              ~ network_tier = "PREMIUM" -> (known after apply)
            }
        }

      ~ scheduling {
            automatic_restart   = true
          ~ on_host_maintenance = "MIGRATE" -> (known after apply)
            preemptible         = false
        }

      ~ service_account {
          ~ email  = "default" -> (known after apply)
            scopes = [
                "https://www.googleapis.com/auth/devstorage.read_only",
                "https://www.googleapis.com/auth/logging.write",
                "https://www.googleapis.com/auth/monitoring.write",
                "https://www.googleapis.com/auth/service.management.readonly",
                "https://www.googleapis.com/auth/servicecontrol",
                "https://www.googleapis.com/auth/trace.append",
            ]
        }
    }

  # google_compute_region_instance_group_manager.asd_c["europe-west1"] will be updated in-place
  ~ resource "google_compute_region_instance_group_manager" "asd_c" {
        base_instance_name        = "default"
        distribution_policy_zones = [
            "europe-west1-b",
            "europe-west1-c",
            "europe-west1-d",
        ]
        fingerprint               = "sgRCc75Md1s="
        id                        = "projects/someproject/regions/europe-west1/instanceGroupManagers/default-europe-west1-asd-c"
        instance_group            = "https://www.googleapis.com/compute/v1/projects/someproject/regions/europe-west1/instanceGroups/default-europe-west1-asd-c"
        name                      = "default-europe-west1-asd-c"
        project                   = "someproject"
        region                    = "europe-west1"
        self_link                 = "https://www.googleapis.com/compute/v1/projects/someproject/regions/europe-west1/instanceGroupManagers/default-europe-west1-asd-c"
        target_pools              = []
        target_size               = 1
        wait_for_instances        = false

        update_policy {
            instance_redistribution_type = "NONE"
            max_surge_fixed              = 3
            max_surge_percent            = 0
            max_unavailable_fixed        = 3
            max_unavailable_percent      = 0
            min_ready_sec                = 480
            minimal_action               = "REPLACE"
            type                         = "PROACTIVE"
        }

      ~ version {
          ~ instance_template = "https://www.googleapis.com/compute/v1/projects/someproject/global/instanceTemplates/default20191128160228194200000004" -> (known after apply)
            name              = "current"
        }
    }

Plan: 4 to add, 1 to change, 4 to destroy.
```

--- a/terraform/evaluate.go
+++ b/terraform/evaluate.go
@@ -648,60 +648,59 @@ func (d *evaluationStateData) getResourceInstancesAll(addr addrs.Resource, rng t
 		for i := 0; i < length; i++ {
 			ty := schema.ImpliedType()
 			key := addrs.IntKey(i)
-			is, exists := rs.Instances[key]
-			if exists && is.Current != nil {
-				instAddr := addr.Instance(key).Absolute(d.ModulePath)
-
-				// Prefer pending value in plan if present. See getResourceInstanceSingle
-				// comment for the rationale.
-				if is.Current.Status == states.ObjectPlanned {
-					if change := d.Evaluator.Changes.GetResourceInstanceChange(instAddr, states.CurrentGen); change != nil {
-						val, err := change.After.Decode(ty)
-						if err != nil {
-							diags = diags.Append(&hcl.Diagnostic{
-								Severity: hcl.DiagError,
-								Summary:  "Invalid resource instance data in plan",
-								Detail:   fmt.Sprintf("Instance %s data could not be decoded from the plan: %s.", instAddr, err),
-								Subject:  &config.DeclRange,
-							})
-							continue
-						}
-						vals[i] = val
-						continue
-					} else {
-						// If the object is in planned status then we should not
-						// get here, since we should've found a pending value
-						// in the plan above instead.
-						diags = diags.Append(&hcl.Diagnostic{
-							Severity: hcl.DiagError,
-							Summary:  "Missing pending object in plan",
-							Detail:   fmt.Sprintf("Instance %s is marked as having a change pending but that change is not recorded in the plan. This is a bug in Terraform; please report it.", instAddr),
-							Subject:  &config.DeclRange,
-						})
-						continue
-					}
-				}
-
-				ios, err := is.Current.Decode(ty)
-				if err != nil {
-					// This shouldn't happen, since by the time we get here
-					// we should've upgraded the state data already.
-					diags = diags.Append(&hcl.Diagnostic{
-						Severity: hcl.DiagError,
-						Summary:  "Invalid resource instance data in state",
-						Detail:   fmt.Sprintf("Instance %s data could not be decoded from the state: %s.", instAddr, err),
-						Subject:  &config.DeclRange,
-					})
-					continue
-				}
-				vals[i] = ios.Value
-			} else {
+			is := rs.Instances[key]
+			if is == nil || is.Current == nil {
 				// There shouldn't normally be "gaps" in our list but we'll
 				// allow it under the assumption that we're in a weird situation
 				// where e.g. someone has run "terraform state mv" to reorder
 				// a list and left a hole behind.
 				vals[i] = cty.UnknownVal(schema.ImpliedType())
+				continue
 			}
+
+			instAddr := addr.Instance(key).Absolute(d.ModulePath)
+
+			if is.Current.Status == states.ObjectPlanned {
+				if change := d.Evaluator.Changes.GetResourceInstanceChange(instAddr, states.CurrentGen); change != nil {
+					val, err := change.After.Decode(ty)
+					if err != nil {
+						diags = diags.Append(&hcl.Diagnostic{
+							Severity: hcl.DiagError,
+							Summary:  "Invalid resource instance data in plan",
+							Detail:   fmt.Sprintf("Instance %s data could not be decoded from the plan: %s.", instAddr, err),
+							Subject:  &config.DeclRange,
+						})
+						continue
+					}
+					vals[i] = val
+					continue
+				} else {
+					// If the object is in planned status then we should not
+					// get here, since we should've found a pending value
+					// in the plan above instead.
+					diags = diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Missing pending object in plan",
+						Detail:   fmt.Sprintf("Instance %s is marked as having a change pending but that change is not recorded in the plan. This is a bug in Terraform; please report it.", instAddr),
+						Subject:  &config.DeclRange,
+					})
+					continue
+				}
+			}
+
+			ios, err := is.Current.Decode(ty)
+			if err != nil {
+				// This shouldn't happen, since by the time we get here
+				// we should've upgraded the state data already.
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid resource instance data in state",
+					Detail:   fmt.Sprintf("Instance %s data could not be decoded from the state: %s.", instAddr, err),
+					Subject:  &config.DeclRange,
+				})
+				continue
+			}
+			vals[i] = ios.Value
 		}
 
 		// We use a tuple rather than a list here because resource schemas may
@@ -715,12 +714,14 @@ func (d *evaluationStateData) getResourceInstancesAll(addr addrs.Resource, rng t
 		vals := make(map[string]cty.Value, len(rs.Instances))
 		for k, is := range rs.Instances {
 			if sk, ok := k.(addrs.StringKey); ok {
+				if is == nil || is.Current == nil {
+					// Assume we're dealing with an instance that hasn't been created yet.
+					vals[string(sk)] = cty.UnknownVal(schema.ImpliedType())
+					continue
+				}
+
 				instAddr := addr.Instance(k).Absolute(d.ModulePath)
 
-				// Prefer pending value in plan if present. See getResourceInstanceSingle
-				// comment for the rationale.
-				// Prefer pending value in plan if present. See getResourceInstanceSingle
-				// comment for the rationale.
 				if is.Current.Status == states.ObjectPlanned {
 					if change := d.Evaluator.Changes.GetResourceInstanceChange(instAddr, states.CurrentGen); change != nil {
 						val, err := change.After.Decode(ty)

